src/modules/job-manager/queue.c

```diff
@@ -774,6 +776,109 @@ void queue_destroy (struct queue *queue)
     }
 }
 
+static int queue_update_cb (flux_plugin_t *p,
+                            const char *topic,
+                            flux_plugin_arg_t *args,
+                            void *arg)
+{
+    struct queue *queue = arg;
+    flux_job_state_t state;
+    const char *name;
+    const char *current_queue = NULL;
+    json_t *constraints = NULL;
+    flux_error_t error;
+    struct jobq *newq;
+
+    if (flux_plugin_arg_unpack (args,
+                                FLUX_PLUGIN_ARG_IN,
+                                "{s:s s:i s:{s:{s:{s?s s?o}}}}",
+                                "value", &name,
+                                "state", &state,
+                                "jobspec",
+                                 "attributes",
+                                  "system",
+                                   "queue", &current_queue,
+                                   "constraints", &constraints) < 0) {
+        flux_jobtap_error (p, args, "plugin args unpack failed");
+        return -1;
+    }
+    if (state == FLUX_JOB_STATE_RUN
+        || state == FLUX_JOB_STATE_CLEANUP) {
+        flux_jobtap_error (p,
+                           args,
+                           "update of queue for running job not supported");
+        return -1;
+    }
+    if (streq (current_queue, name)) {
+        flux_jobtap_error (p,
+                           args,
+                           "job queue is already set to %s",
+                           name);
+        return -1;
+    }
+    if (!(newq = queue_lookup (queue, name, &error))) {
+        flux_jobtap_error (p, args, "%s", error.text);
+        return -1;
+    }
+    if (!newq->enable) {
+        flux_jobtap_error (p,
+                           args,
+                           "queue %s is currently disabled",
+                           name);
+        return -1;
+    }
+    /*  Constraints must match current queue exactly since they will be
+     *  overwritten:
+     */
+    if (constraints) {
+        json_t *expected;
+        struct jobq *q;
+        if (!(q = queue_lookup (queue, current_queue, &error))
+            || !q->requires
+            || !(expected = json_pack ("{s:O}", "properties", q->requires))) {
+            flux_jobtap_error (p,
+                               args,
+                               "failed to get constraints for current queue");
+            return -1;
+        }
+        if (!json_equal (constraints, expected)) {
+            flux_jobtap_error (p,
+                               args,
+                               "job appears to have non-queue constraints, "
+                               "unable to update queue to %s",
+                               name);
+            json_decref (expected);
+            return -1;
+        }
+        json_decref (expected);
```

Comment:
I started pulling this out into its own function for more clarity, then realized that there are some bugs here when a queue does not have a configured constraint. This will require a bit more code and tests, so working on that now.